REGRESSION(257749@main): Copying table in Confluence on Safari doesn't copy header rows

rniwa opened the pull request:
#### 6f132509a670a1b3e8ef9aa552736926b456e8ee
<pre>
REGRESSION(257749@main): Copying table in Confluence on Safari doesn&apos;t copy header rows
<a href="https://bugs.webkit.org/show_bug.cgi?id=259091">https://bugs.webkit.org/show_bug.cgi?id=259091</a>

Reviewed by Wenson Hsieh.

Add a Confluence specific quirk to copy `-webkit-user-select: none` contents.

* LayoutTests/editing/pasteboard/copy-content-with-user-select-none-quirks-expected.txt: Added.
* LayoutTests/editing/pasteboard/copy-content-with-user-select-none-quirks.html: Added.
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::selectedText const):
(WebCore::Editor::selectedTextForDataTransfer const):
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::HTMLConverter):
(HTMLConverter::_enterElement):
(HTMLConverter::_processText):
* Source/WebCore/editing/markup.cpp:
(WebCore::StyledMarkupAccumulator::StyledMarkupAccumulator):
* Source/WebCore/html/HTMLMetaElement.cpp:
(WebCore::HTMLMetaElement::process):
* Source/WebCore/page/DOMSelection.cpp:
(WebCore::DOMSelection::toString const):
* Source/WebCore/page/Quirks.h:
(WebCore::Quirks::needsToCopyUserSelectNoneQuirk const):
(WebCore::Quirks::setNeedsToCopyUserSelectNoneQuirk):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CopyRTF.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/265939@main">https://commits.webkit.org/265939@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f905973ab0e74a137cfbc3fd5c8d84b7ecd4822a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12328 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12664 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12974 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14071 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11863 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12384 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15089 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12681 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14545 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12492 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13263 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10429 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14492 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10546 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11171 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18279 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11629 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11336 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14518 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11824 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9762 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11050 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11045 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15380 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1377 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11678 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebCore/page/Quirks.h:
```diff
@@ -170,6 +170,9 @@ class Quirks {
     void setNeedsConfigurableIndexedPropertiesQuirk() { m_needsConfigurableIndexedPropertiesQuirk = true; }
     bool needsConfigurableIndexedPropertiesQuirk() const;
 
+    bool needsToCopyUserSelectNoneQuirk() const { return m_needsToCopyUserSelectNoneQuirk; }
```
Could you add the reference to the bug in Webkit or radar? So it's easier to maintain when we need to remove it. Thanks.

karlcow replied:
As an example you can check the modifications on Quirks.cpp

rniwa replied:
Sure, although that's what blame is for.

karlcow reviewed (changes requested):
quick review just for the references parts.

rniwa committed:
REGRESSION(257749@main): Copying table in Confluence on Safari doesn't copy header rows
https://bugs.webkit.org/show_bug.cgi?id=259091

Reviewed by Wenson Hsieh.

Add a Confluence specific quirk to copy `-webkit-user-select: none` contents.

* LayoutTests/editing/pasteboard/copy-content-with-user-select-none-quirks-expected.txt: Added.
* LayoutTests/editing/pasteboard/copy-content-with-user-select-none-quirks.html: Added.
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::selectedText const):
(WebCore::Editor::selectedTextForDataTransfer const):
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::HTMLConverter):
(HTMLConverter::_enterElement):
(HTMLConverter::_processText):
* Source/WebCore/editing/markup.cpp:
(WebCore::StyledMarkupAccumulator::StyledMarkupAccumulator):
* Source/WebCore/html/HTMLMetaElement.cpp:
(WebCore::HTMLMetaElement::process):
* Source/WebCore/page/DOMSelection.cpp:
(WebCore::DOMSelection::toString const):
* Source/WebCore/page/Quirks.h:
(WebCore::Quirks::needsToCopyUserSelectNoneQuirk const):
(WebCore::Quirks::setNeedsToCopyUserSelectNoneQuirk):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CopyRTF.mm:
(TEST):

Canonical link: https://commits.webkit.org/265939@main
